src/internal/storage/fileset/index/reader.go

```diff
@@ -190,7 +200,100 @@ func (lr *levelReader) next() error {
 	if err := lr.parent.Read(lr.idx); err != nil {
 		return errors.EnsureStack(err)
 	}
+	if lr.idx.File != nil {
+		return io.EOF
+	}
 	r := lr.r.chunks.NewReader(lr.ctx, []*chunk.DataRef{lr.idx.Range.ChunkRef})
 	lr.buf.Reset()
 	return r.Get(lr.buf)
 }
+
+// TODO: Making this configurable based on the number of workers available might be the right long term strategy.
+const numShards = 10
+
+// ShardCallback is a callback that returns a path range for each shard.
+type ShardCallback = func(*PathRange) error
+
+func (r *Reader) Shard(ctx context.Context, cb ShardCallback) error {
+	if r.topIdx == nil || r.topIdx.File != nil {
+		return cb(&PathRange{})
+	}
+	idxs := []*Index{r.topIdx}
+Loop:
+	for len(idxs) < numShards {
+		nextIdxs, err := r.nextLevel(ctx, idxs)
+		if err != nil {
+			return err
+		}
+		// TODO: Unbalanced shards can result with files & ranges at the same level.
+		for _, idx := range nextIdxs {
+			if idx.File != nil {
+				break Loop
+			}
+		}
+		idxs = nextIdxs
+	}
+	for _, shard := range resizeShards(createShards(idxs), numShards) {
+		if err := cb(shard); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func createShards(idxs []*Index) []*PathRange {
+	var shards []*PathRange
+	for i, idx := range idxs {
+		pathRange := &PathRange{}
+		if i != 0 {
+			pathRange.Lower = idx.Path
+		}
+		if i != len(idxs)-1 {
+			pathRange.Upper = idxs[i+1].Path
+		}
+		shards = append(shards, pathRange)
+	}
+	return shards
+}
+
+func resizeShards(shards []*PathRange, maxShards int) []*PathRange {
+	if len(shards) <= maxShards {
+		return shards
+	}
+	var newShards []*PathRange
+	size := len(shards) / maxShards
+	shard := &PathRange{}
+	for i := size; i < len(shards); i += size {
+		shard.Upper = shards[i].Lower
+		newShards = append(newShards, shard)
+		shard = &PathRange{
+			Lower: shards[i].Lower,
+		}
+	}
+	return append(newShards, shard)
+}
+
+func (r *Reader) nextLevel(ctx context.Context, idxs []*Index) ([]*Index, error) {
+	var dataRefs []*chunk.DataRef
+	for _, idx := range idxs {
+		dataRefs = append(dataRefs, idx.Range.ChunkRef)
+	}
+	cr := r.chunks.NewReader(ctx, dataRefs)
+	buf := &bytes.Buffer{}
+	if err := cr.Get(buf); err != nil {
```

Comment:
It could be concerning if it was used more generally, but I don't anticipate that. For the current usage, we will never get low enough for this to be a problem since we stop when we get past numShards path ranges. I don't think we should preemptively make this more generally scalable.